lighthouse-core/gather/computed/metrics/lantern-metric.js

```diff
@@ -0,0 +1,126 @@
+const Simulator = require('../../../lib/dependency-graph/simulator/simulator');
+const WebInspector = require('../../../lib/web-inspector');
+
+class LanternMetricArtifact extends ComputedArtifact {
+  /**
+   * @param {!Node} dependencyGraph
+   * @param {function(NetworkNode):boolean=} condition
+   * @return {!Set<string>}
+   */
+  static getScriptUrls(dependencyGraph, condition) {
+    const scriptUrls = new Set();
+
+    dependencyGraph.traverse(node => {
+      if (node.type === Node.TYPES.CPU) return;
+      const asNetworkNode = /** @type {NetworkNode} */ (node);
+      if (asNetworkNode.record._resourceType !== WebInspector.resourceTypes.Script) return;
+      if (condition && !condition(asNetworkNode)) return;
+      scriptUrls.add(asNetworkNode.record.url);
+    });
+
+    return scriptUrls;
+  }
+
+  /**
+   * @return {LH.Gatherer.Simulation.MetricCoefficients}
+   */
+  get COEFFICIENTS() {
+    throw new Error('COEFFICIENTS unimplemented!');
+  }
+
+  /**
+   * @param {!Node} dependencyGraph
+   * @param {LH.Gatherer.Artifact.TraceOfTab} traceOfTab
+   * @return {!Node}
+   */
+  getOptimisticGraph(dependencyGraph, traceOfTab) { // eslint-disable-line no-unused-vars
+    throw new Error('Optimistic graph unimplemented!');
+  }
+
+  /**
+   * @param {!Node} dependencyGraph
+   * @param {LH.Gatherer.Artifact.TraceOfTab} traceOfTab
+   * @return {!Node}
+   */
+  getPessimisticGraph(dependencyGraph, traceOfTab) { // eslint-disable-line no-unused-vars
+    throw new Error('Pessmistic graph unimplemented!');
+  }
+
+  /**
+   * @param {LH.Gatherer.Simulation.Result} simulationResult
+   * @param {*} extras
+   * @return {number}
+   */
+  getEstimateFromSimulation(simulationResult, extras) { // eslint-disable-line no-unused-vars
+    return simulationResult.timeInMs;
+  }
+
+  /**
+   * @param {{trace: Object, devtoolsLog: Object}} data
+   * @param {Object} artifacts
+   * @param {any=} extras
+   * @return {Promise<LH.Gatherer.Artifact.LanternMetric>}
+   */
+  async computeMetricWithGraphs(data, artifacts, extras) {
+    const {trace, devtoolsLog} = data;
+    const graph = await artifacts.requestPageDependencyGraph({trace, devtoolsLog});
+    const traceOfTab = await artifacts.requestTraceOfTab(trace);
+    const networkAnalysis = await artifacts.requestNetworkAnalysis(devtoolsLog);
+
+    const optimisticGraph = this.getOptimisticGraph(graph, traceOfTab);
+    const pessimisticGraph = this.getPessimisticGraph(graph, traceOfTab);
+
+    // TODO(phulce): use rtt and throughput from config.settings instead of defaults
+    const options = {
+      additionalRttByOrigin: networkAnalysis.additionalRttByOrigin,
+      serverResponseTimeByOrigin: networkAnalysis.serverResponseTimeByOrigin,
+    };
+
+    const optimisticEstimate = new Simulator(optimisticGraph, options).simulate();
+    const pessimisticEstimate = new Simulator(pessimisticGraph, options).simulate();
+
+    optimisticEstimate.timeInMs = this.getEstimateFromSimulation(
```

Comment:
this bit still seems weird to me. here's how i'm reading it:

1. simulator, plz simulate with this graph. 
2. simulation done here's my estimate and some nodeTiming internals
3. okay thx 
   - but if this is the CI metric, i'm gonna do some extra work where I extract the lastLongTaskEndTime from the iternals and also take the lantern FMP number, and do some `Math.max` with those. and then i'm gonna ignore the provided estimate and instead use this one.

Is that right? 

I'm thinking one alternative a sort of `minimumEstimateTime(nodeTiming){ }` method that's either provided to the simulator/simulate or something?   Don't know if that makes sense though..


Are there any clients of `simulate()` now that don't override `timeInMs` with their own value?

Reply:
I think this is a misunderstanding based on my laziness :) Simulator has no concept of a metric at all, and that's a good thing we'd like to keep; it just computes timings of things. 

Metrics take these timings and return estimates, in the current case 1 optimistic, 1 pessimistic.

It just so happens that 2 of our 3 current metrics use time of the last network resource (what simulator outputs) as their estimate without any additional logic. Rather than add a `mapSimulatorOutputToEstimateOutput` I abused the existing return object and allow the metric to mutate it if it wants.

Given the amount of confusion this causes and the fact that we're about to be in a situation where only 2/5 metrics take simulator output without touching it, I think it makes sense to just go the `mapSimulatorOutputToEstimateOutput` route, WDYT? Does this make more sense?

Reply:
@brendankenny since paul's OOO for awhile WDYT of the above? :)

Reply:
SGTM